runtime/lua/vim/lsp/util.lua

```diff
@@ -440,16 +421,35 @@ function M.apply_text_edits(text_edits, bufnr, offset_encoding)
 
   -- Apply text edits.
   local is_cursor_fixed = false
+  local has_eol_text_edit = false
   for _, text_edit in ipairs(text_edits) do
+    -- Convert from LSP style ranges to Neovim style ranges.
     local e = {
       start_row = text_edit.range.start.line,
-      start_col = get_line_byte_from_position(bufnr, text_edit.range.start),
+      start_col = get_line_byte_from_position(bufnr, text_edit.range.start, offset_encoding),
       end_row = text_edit.range['end'].line,
-      end_col  = get_line_byte_from_position(bufnr, text_edit.range['end']),
+      end_col  = get_line_byte_from_position(bufnr, text_edit.range['end'], offset_encoding),
       text = vim.split(text_edit.newText, '\n', true),
     }
+
+    -- Some LSP servers may return +1 range of the buffer content but nvim_buf_set_text can't accept it so we should fix it here.
+    local max = vim.api.nvim_buf_line_count(bufnr)
+    local len = #(vim.api.nvim_buf_get_lines(bufnr, -2, -1, false)[1] or '')
```

Comment:
should we use get_line here instead? Also this could be guarded under the conditional 
`if max <= e.start_row or max <= e.end_row then` to avoid querying the buf.